zerver/tests/test_previewable.py

```diff
@@ -0,0 +1,140 @@
+        self.assertEqual(result["pull_request"], {"merged_at": None})
+
+    def test_not_previewable_url(self) -> None:
+        result = self.client_post(
+            "/json/previewable", {"url": "https://github.com/zulip/xyz/zulip/issues/1"}
+        )
+        self.assert_json_error(result, msg="URL is not previewable.")
+
+    @responses.activate()
+    def test_invalid_previewable_url(self) -> None:
+        response_data = (
+            {
+                "message": "Not Found",
+                "documentation_url": "https://docs.github.com/rest/reference/issues#get-an-issue",
+            },
+        )
+        api_url = "https://api.github.com/repos/zulip/zulip/issues/2"
+        responses.add(
+            responses.GET,
+            api_url,
+            json=response_data,
+            status=404,
+        )
+        with self.assertLogs("zerver.lib.github", level="ERROR") as error_log:
+            result = self.client_post(
+                "/json/previewable", {"url": "https://github.com/zulip/zulip/issues/2"}
+            )
+            self.assert_json_error(result, "Invalid or expired URL.")
+            self.assertIn(
+                f"ERROR:zerver.lib.github:Unable to fetch data from GitHub {api_url}",
+                error_log.output[0],
+            )
+
+    @responses.activate()
+    def test_github_internal_server_error(self) -> None:
+        api_url = "https://api.github.com/repos/zulip/zulip/issues/2"
+        responses.add(
+            responses.GET,
+            api_url,
+            status=500,
+        )
+        with self.assertLogs("zerver.lib.github", level="ERROR") as error_log:
+            result = self.client_post(
+                "/json/previewable", {"url": "https://github.com/zulip/zulip/issues/2"}
+            )
+            self.assert_json_error(result, "Unable to fetch data from Github.")
+            self.assertIn(
+                f"ERROR:zerver.lib.github:Unable to fetch data from GitHub {api_url}",
+                error_log.output[0],
+            )
+
+    @responses.activate()
+    def test_cache(self) -> None:
+        response_data = {
+            "title": "Fix Markdown formatting of code blocks & snippets",
+            "user": {"login": "thomasboyt"},
+            "state": "open",
+            "locked": False,
+            "state_reason": None,
+        }
+        responses.add(
+            responses.GET, "https://api.github.com/repos/zulip/zulip/issues/1", json=response_data
+        )
+        self.client_post("/json/previewable", {"url": "https://github.com/zulip/zulip/issues/1"})
+        response_data = {
+            "title": "Fix Markdown formatting of code blocks & snippets",
+            "user": {"login": "thomasboyt"},
+            "state": "closed",
+            "locked": False,
+            "state_reason": None,
+        }
+        responses.add(
+            responses.GET, "https://api.github.com/repos/zulip/zulip/issues/1", json=response_data
+        )
```

Comment:
Shouldn't the cache mean that we _don't_ need this response?

Reply:
Ooh, I missed that, will address this in next commit.

Reply:
If the cache is working, then it *shouldn't* see the new data in the response, because it should return the cached data from the first response, and not make a second request.

Reply:
Okay, I will update this to like we will first add a mock response than request for data than remove the mock response and than again fetch the data and assert.